### PR TITLE
Holes

### PR DESCRIPTION
**Pull request by Uppun**
I think it's all polished up now. lmk what you think

### CODE REVIEW

File: src/Cell.js [hunk 1]
```diff
@@ -1,24 +1,26 @@
+const neighborOffsets = [
+    {coords: [-1, -1], cost: 1.5},
+    {coords: [0, -1], cost: 1},
+    {coords: [1, -1], cost: 1.5},
+    {coords: [-1, 0], cost: 1},
+    {coords: [1, 0], cost: 1},
+    {coords: [-1, 1], cost: 1.5},
+    {coords: [0, 1], cost: 1},
+    {coords: [1, 1], cost: 1.5}];
```
Comment (meishuu) [inline src/Cell.js]: nitpinion: if `[` has newline, i think `]` should too

File: src/Heap.js [hunk 2]
```diff
@@ -1,78 +1,80 @@
...
         this.minHeapify(0);
 
         return node; 
     }
 
-    setFscore(node, fScore) {
-        const nodeIndex = this._nodeToIndexMap.get(node);
-        if (nodeIndex) {
-            this.heap[nodeIndex].f = fScore;
+    setFscore(node, key) {
```
Comment (meishuu) [inline src/Heap.js]: still references "f score" 🙁

File: src/Heap.js [hunk 3]
```diff
@@ -1,78 +1,80 @@
...
         const lastParent = Math.floor((this.heap.length - 2) / 2);
 
-        for(let i = lastParent; i >= 0; i--) {
+        for (let i = lastParent; i >= 0; i--) {
             this.minHeapify(i);
         }
     }
 
-    pushNode(length) {
+    heapCheck(length) {
```
Comment (meishuu) [inline src/Heap.js]: nitpinion: pretty much all of your methods in here are named in `[verb][Noun?]` form... except this one

File: src/Heap.js [hunk 4]
```diff
@@ -1,78 +1,80 @@
...
         if (length > 0) {
-            const position = Math.floor((length-2)/2);
+            const position = Math.floor((length - 2) / 2);
             this.minHeapify(position);
-            this.pushNode(position);
+            this.heapCheck(position);
```
Comment (meishuu) [inline src/Heap.js]: you could theoretically just make this a while loop
recursion is not-great b/c most js engines don't do [tco](http://2ality.com/2015/06/tail-call-optimization.html) yet

File: src/Heap.js [hunk 5]
```diff
@@ -1,78 +1,80 @@
...
         }
     }
 
-    insert(node) {
-        this.heap.push(node);
-        this._nodeToIndexMap.set(node.node, this.heap.length - 1);
-        this.pushNode(this.heap.length - 1);
+    insert(node, f) {
```
Comment (meishuu) [inline src/Heap.js]: still an "f (score)" 🙁

File: src/components/GridComponent.js [hunk 6]
```diff
@@ -40,37 +46,55 @@ class PathfinderGrid extends Component {
...
-    handleGenerateClick = () => {
+    handleGenerateClick() {
         GridActions.generate();
     }
 
+    handleCellClick = (x, y) => {
+        const {stage} = this.state;
+        
+        switch (stage) {
+            case "STARTP": {
```
Comment (meishuu) [inline src/components/GridComponent.js]: nit: single quotes

File: src/components/GridComponent.js [hunk 7]
```diff
@@ -40,37 +46,55 @@ class PathfinderGrid extends Component {
...
-                                    return (<Cell status={cell.passable} key={columnIndex} type={'player'} stage={stage} position={{x: cell.x, y: cell.y}} />);
-                                }
-
-                                if (end.y === rowIndex && end.x === columnIndex) {
-                                    return (<Cell status={cell.passable} key={columnIndex} type={'end'} stage={stage} position={{x: cell.x, y: cell.y}} />);
-                                }
-                            
-                                return (<Cell status={cell.passable} key={columnIndex} type={'normal'} stage={stage} position={{x: cell.x, y: cell.y}} />);
+                                const type = player.y === rowIndex && player.x === columnIndex ? "player" :
+                                             end.y === rowIndex && end.x === columnIndex ? "end" : "normal";
```
Comment (meishuu) [inline src/components/GridComponent.js]: nit: single quotes

File: src/stores/A_store.js [hunk 8]
```diff
@@ -4,65 +4,64 @@ import ActionTypes from '../actions/ActionTypes';
 import Grid from '../Grid';
 import Astar from '../Astar';
 import mapGenerator from '../mapGenerator';
+import heuristic from '../heuristic';
 
 const SIZE = 50; 
 
-class PathfindStore extends ReduceStore {
+class A_Store extends ReduceStore {
```
Comment (meishuu) [inline src/stores/A_store.js]: filename is `A_store`

File: src/stores/A_store.js [hunk 9]
```diff
@@ -4,65 +4,64 @@ import ActionTypes from '../actions/ActionTypes';
...
     constructor() {
         super(Dispatcher);
     }
 
     getInitialState() {
-        const pathGrid = new Grid();
-        return {pathGrid, stage: 'STARTP', player: {x: -1, y: -1}, end: {x: -2, y: -2}};
+        const pathGrid = new Grid(SIZE, SIZE);
+        return {pathGrid,
```
Comment (meishuu) [inline src/stores/A_store.js]: nitpinion:
```js
return {
    pathGrid: new Grid(SIZE, SIZE),
    stage: 'STARTP',
    player: {x: -1, y: -1},
    end: {x: -2, y: -2},
};
```

File: src/stores/A_store.js [hunk 10]
```diff
@@ -4,65 +4,64 @@ import ActionTypes from '../actions/ActionTypes';
...
-                            pathGrid.grid[y][x].passable = !grid[y][x].passable;
-                        }
-                        else {
-                            pathGrid.grid[y][x].passable = grid[y][x].passable;
-                        }
+                const {x, y} = action;
+
+                for (let i = 0; i < SIZE; i++) {
+                    for (let j = 0; j < SIZE; j++) {
+                        pathGrid.grid[i][j].passable = grid[i][j].passable;
```
Comment (meishuu) [inline src/stores/A_store.js]: upon further reflection i feel like we should just support some sort of `Grid#clone()` method

File: src/stores/A_store.js [hunk 11]
```diff
@@ -4,65 +4,64 @@ import ActionTypes from '../actions/ActionTypes';
...
                 path.shift();
                 
                 return {...state, path, stage: 'STEP'};
             }
 
             case ActionTypes.RESET: {
-                let pathGrid = new Grid();
+                const pathGrid = new Grid(SIZE, SIZE);
 
                 return {pathGrid, player: {x: -1, y: -1}, end: {x: -2, y: -2}, stage: 'STARTP', path: ''};
```
Comment (meishuu) [inline src/stores/A_store.js]: this is also still one line 🤔 
can also get rid of the variable here too if we do it like in `getInitialState`

File: src/stores/A_store.js [hunk 12]
```diff
@@ -71,13 +70,7 @@ class PathfindStore extends ReduceStore {
...
                 const nextLocation = path.shift();
 
-                let stage;
-                if (path.length > 0) {
-                    stage = "STEP";
-                }
-                else {
-                    stage = "RESET";
-                }
+                const stage = path.length > 0 ? "STEP" : "RESET";
```
Comment (meishuu) [inline src/stores/A_store.js]: single quotes

File: src/components/GridComponent.js [hunk 13]
```diff
@@ -6,6 +6,12 @@ import GridActions from '../actions/GridActions';
 
 import '../main.css';
 
+const STAGE_LABELS = {
+    WALL: 'Pathfind',
```
Comment (meishuu) [inline src/components/GridComponent.js]: what is this indenting 🤔 

**edit:** meant for line 13 oop

File: src/actions/ActionTypes.js [hunk 14]
```diff
@@ -1,7 +1,7 @@
 export default {
     STARTP: 'STARTP',
     ENDP: 'ENDP',
-    PASS: 'PASS',
+    PASS: 'WALL',
```
Comment (meishuu) [inline src/actions/ActionTypes.js]: should be changing `PASS` too then

File: src/Heap.js [hunk 15]
```diff
@@ -1,78 +1,80 @@
...
+        }
+
+        let smallest = position;
+        const leftChild = 2 * position + 1;
+        const rightChild = 2 * position + 2;
             
-            if (leftChild <= this.heap.length - 1 && this.fCompare(this.heap[leftChild], this.heap[position]) < 0) {
-                smallest = leftChild;
-            }
+        if (leftChild <= this.heap.length - 1 && this.compare(this.heap[leftChild], this.heap[position]) < 0) {
```
Comment (meishuu) [inline src/Heap.js]: instead of `<= len - 1` should just `< len`

File: src/Grid.js [hunk 16]
```diff
@@ -1,24 +1,21 @@
 import Cell from './Cell';
 
 export default class Grid {
-    constructor() {
-        this.height = 50;
-        this.width = 50;
-        this.grid = this.generateGrid();
+    constructor(height, width) {
```
Comment (meishuu) [inline src/Grid.js]: 100% opinion: (width, height) feels much more common than (height, width)

File: src/Heap.js [hunk 17]
```diff
@@ -1,78 +1,80 @@
 export default class Heap {
     heap = [];
-    _nodeToIndexMap = new Map();
+    _nodeIndicies = new Map();
 
-    extractRoot(){
+    getLength() {
```
Comment (meishuu) [inline src/Heap.js]: close – but there's a way to implement these such that they look like actual properties and not methods

File: src/components/GridComponent.js (hunk 13 above)
Comment (Uppun) [reply src/components/GridComponent.js]: If you are on the wall placing stage, the button should display "Pathfind." Can't really pathfind if you haven't placed a start and end goal yet.

File: src/Cell.js (hunk 1 above)
Comment (Uppun) [reply src/Cell.js]: 🐇

File: src/components/GridComponent.js (hunk 13 above)
Comment (Uppun) [reply src/components/GridComponent.js]: oh i misread indenting as intending

File: src/Heap.js (hunk 2 above)
Comment (Uppun) [reply src/Heap.js]: 🐇

File: src/Heap.js (hunk 3 above)
Comment (Uppun) [reply src/Heap.js]: 🐇

File: src/Heap.js (hunk 17 above)
Comment (Uppun) [reply src/Heap.js]: 🐇

File: src/Heap.js (hunk 4 above)
Comment (Uppun) [reply src/Heap.js]: 🐇

File: src/Heap.js (hunk 5 above)
Comment (Uppun) [reply src/Heap.js]: 🐇

File: src/components/GridComponent.js (hunk 6 above)
Comment (Uppun) [reply src/components/GridComponent.js]: 🐇

File: src/components/GridComponent.js (hunk 7 above)
Comment (Uppun) [reply src/components/GridComponent.js]: 🐇

File: src/stores/A_store.js (hunk 8 above)
Comment (Uppun) [reply src/stores/A_store.js]: 🐇

File: src/stores/A_store.js (hunk 9 above)
Comment (Uppun) [reply src/stores/A_store.js]: 🐇

File: src/stores/A_store.js (hunk 10 above)
Comment (Uppun) [reply src/stores/A_store.js]: 🐇

File: src/stores/A_store.js (hunk 11 above)
Comment (Uppun) [reply src/stores/A_store.js]: 🐇

File: src/stores/A_store.js (hunk 12 above)
Comment (Uppun) [reply src/stores/A_store.js]: 🐇

File: src/components/GridComponent.js (hunk 13 above)
Comment (Uppun) [reply src/components/GridComponent.js]: 🐇

File: src/actions/ActionTypes.js (hunk 14 above)
Comment (Uppun) [reply src/actions/ActionTypes.js]: 🐇

File: src/Heap.js (hunk 15 above)
Comment (Uppun) [reply src/Heap.js]: 🐇

File: src/Grid.js (hunk 16 above)
Comment (Uppun) [reply src/Grid.js]: 🐇 😡

File: src/Heap.js [hunk 18]
```diff
@@ -1,78 +1,79 @@
...
         return node; 
     }
 
-    setFscore(node, fScore) {
-        const nodeIndex = this._nodeToIndexMap.get(node);
-        if (nodeIndex) {
-            this.heap[nodeIndex].f = fScore;
+    setKey(node, key) {
+        const nodeIndex = this._nodeIndicies.get(node);
+        if (nodeIndex && nodeIndex !== 0) {
```
Comment (meishuu) [inline src/Heap.js]: why do we not want to update the root?

File: src/Heap.js [hunk 19]
```diff
@@ -1,78 +1,79 @@
 export default class Heap {
     heap = [];
-    _nodeToIndexMap = new Map();
+    _nodeIndicies = new Map();
```
Comment (meishuu) [inline src/Heap.js]: `indices`

File: src/Grid.js [hunk 20]
```diff
@@ -1,24 +1,33 @@
 import Cell from './Cell';
 
 export default class Grid {
-    constructor() {
-        this.height = 50;
-        this.width = 50;
-        this.grid = this.generateGrid();
+    constructor(width, height) {
+        this.height = height;
```
Comment (meishuu) [inline src/Grid.js]: <sub>turbo nit: we should match arg order</sub>

File: src/Cell.js [hunk 21]
```diff
@@ -1,24 +1,27 @@
+const neighborOffsets = [
+    {coords: [-1, -1], cost: 1.5},
+    {coords: [0, -1], cost: 1},
+    {coords: [1, -1], cost: 1.5},
+    {coords: [-1, 0], cost: 1},
+    {coords: [1, 0], cost: 1},
+    {coords: [-1, 1], cost: 1.5},
+    {coords: [0, 1], cost: 1},
+    {coords: [1, 1], cost: 1.5}
+    ];
```
Comment (meishuu) [inline src/Cell.js]: no indent

File: src/components/GridComponent.js [hunk 22]
```diff
@@ -40,37 +45,55 @@ class PathfinderGrid extends Component {
...
-                                    return (<Cell status={cell.passable} key={columnIndex} type={'player'} stage={stage} position={{x: cell.x, y: cell.y}} />);
-                                }
-
-                                if (end.y === rowIndex && end.x === columnIndex) {
-                                    return (<Cell status={cell.passable} key={columnIndex} type={'end'} stage={stage} position={{x: cell.x, y: cell.y}} />);
-                                }
-                            
-                                return (<Cell status={cell.passable} key={columnIndex} type={'normal'} stage={stage} position={{x: cell.x, y: cell.y}} />);
+                                const type = player.y === rowIndex && player.x === columnIndex ? 'player' :
+                                             end.y === rowIndex && end.x === columnIndex ? 'end' : 'normal';
```
Comment (meishuu) [inline src/components/GridComponent.js]: aite in general [never do more than one ternary conditional](https://eslint.org/docs/rules/no-nested-ternary) (b/c you guessed it, readability)
i kinda liked the layout before

my approach to this would be like
```js
let type = 'normal';
if (player.y === rowIndex && player.x === columnIndex) {
  type = 'player';
} else if (end.y === rowIndex && end.x === columnIndex) {
  type = 'end';
}
```

it's still not the greatest thing to read but the flow is 300% more immediately obvious
(alternatively you can *not* initialize it and put the `normal` in a catch-all `else` at the end)

File: src/components/GridComponent.js [hunk 23]
```diff
@@ -40,37 +45,55 @@ class PathfinderGrid extends Component {
...
+                                return (<Cell passable={cell.passable} key={columnIndex} type={type} handleClick={this.handleCellClick} x={cell.x} y={cell.y} />);
                             })}
                         </div>)}
                 </div>
-                {stageButtonVisibility ? <button onClick={this.handleClick}>{stageLabel}</button> : ''}
+                {stageLabel ? <button onClick={this.handleClick}>{stageLabel}</button> : null}
                 {canGenerate ? <button onClick={this.handleGenerateClick}>Generate</button> : ''}
```
Comment (meishuu) [inline src/components/GridComponent.js]: mayb `null` this too instead of `''` while we're here

File: src/Heap.js (hunk 19 above)
Comment (Uppun) [reply src/Heap.js]: 🐇

File: src/Grid.js (hunk 20 above)
Comment (Uppun) [reply src/Grid.js]: 🐇 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡 😡

File: src/Cell.js (hunk 21 above)
Comment (Uppun) [reply src/Cell.js]: 🐇

File: src/components/GridComponent.js (hunk 22 above)
Comment (Uppun) [reply src/components/GridComponent.js]: 🐇

File: src/components/GridComponent.js (hunk 23 above)
Comment (Uppun) [reply src/components/GridComponent.js]: 🐇

File: src/Heap.js (hunk 18 above)
Comment (Uppun) [reply src/Heap.js]: I don't actually remember why I guess I'll test it without

Comment (Uppun) [reply src/Heap.js]: Logically speaking it shouldn't matter and I can't remember if there was some edge case this would be impacting.

**Review by meishuu — APPROVED** [review]
let do
